control/autoware_collision_detector/src/node.cpp

```diff
@@ -0,0 +1,523 @@
+// Copyright 2024 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "autoware/collision_detector/node.hpp"
+
+#include <autoware/universe_utils/geometry/geometry.hpp>
+#include <autoware/universe_utils/ros/uuid_helper.hpp>
+#include <tf2_eigen/tf2_eigen.hpp>
+
+#include <boost/assert.hpp>
+#include <boost/assign/list_of.hpp>
+#include <boost/format.hpp>
+#include <boost/geometry.hpp>
+#include <boost/geometry/geometries/linestring.hpp>
+#include <boost/geometry/geometries/point_xy.hpp>
+
+#include <pcl/common/transforms.h>
+#include <pcl/point_cloud.h>
+#include <pcl_conversions/pcl_conversions.h>
+
+#define EIGEN_MPL2_ONLY
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+
+namespace autoware::collision_detector
+{
+namespace bg = boost::geometry;
+using Point2d = bg::model::d2::point_xy<double>;
+using Polygon2d = bg::model::polygon<Point2d>;
+using autoware::universe_utils::createPoint;
+using autoware::universe_utils::pose2transform;
+
+namespace
+{
+
+geometry_msgs::msg::Point32 createPoint32(const double x, const double y, const double z)
+{
+  geometry_msgs::msg::Point32 p;
+  p.x = x;
+  p.y = y;
+  p.z = z;
+  return p;
+}
+
+Polygon2d createObjPolygon(
+  const geometry_msgs::msg::Pose & pose, const geometry_msgs::msg::Polygon & footprint)
+{
+  geometry_msgs::msg::Polygon transformed_polygon{};
+  geometry_msgs::msg::TransformStamped geometry_tf{};
+  geometry_tf.transform = pose2transform(pose);
+  tf2::doTransform(footprint, transformed_polygon, geometry_tf);
+
+  Polygon2d object_polygon;
+  for (const auto & p : transformed_polygon.points) {
+    object_polygon.outer().push_back(Point2d(p.x, p.y));
+  }
+
+  bg::correct(object_polygon);
+
+  return object_polygon;
+}
+
+Polygon2d createObjPolygon(
+  const geometry_msgs::msg::Pose & pose, const geometry_msgs::msg::Vector3 & size)
+{
+  const double & length_m = size.x / 2.0;
+  const double & width_m = size.y / 2.0;
```

Comment:
Just curious, why are we using references here?

Reply:
Thank you for your review. No reference is needed here, so I've fixed it in cfafc53a691189bb9c47f2ecbf0b4fa285485bbb